src/instrumentation/podmutator.go

```diff
@@ -35,150 +45,250 @@ var (
 		logger.Error(err, "failed to select a New Relic Instrumentation instance for this pod")
 		return pod, err
 	}
-	insts.Java = inst
+	if len(instCandidates) == 0 {
+		logger.Info("no New Relic Instrumentation instance for this Pod")
+		return pod, errNoInstancesAvailable
+	}
 
-	if inst, err = pm.getInstrumentationInstance(ctx, ns, pod, annotationInjectNodeJS); err != nil {
-		// we still allow the pod to be created, but we log a message to the operator's logs
-		logger.Error(err, "failed to select a New Relic Instrumentation instance for this pod")
-		return pod, err
+	instrumentations, err := pm.getLanguageInstrumentationsFn(instCandidates)
+	if err != nil {
+		if errors.Is(err, errMultipleInstancesPossible) {
+			pm.logger.Info("too many New Relic Instrumentation instances for this Pod.  only 1 allowed")
+			return pod, nil
+		} else {
+			return pod, err
+		}
 	}
-	insts.NodeJS = inst
+	mutatedPod := pm.sdkInjector.Inject(ctx, instrumentations, ns, pod)
 
-	if inst, err = pm.getInstrumentationInstance(ctx, ns, pod, annotationInjectPython); err != nil {
-		// we still allow the pod to be created, but we log a message to the operator's logs
-		logger.Error(err, "failed to select a New Relic Instrumentation instance for this pod")
-		return pod, err
+	if licenseKeySecret, err := pm.getLicenseKeySecretFn(instCandidates); err != nil {
+		logger.Error(err, "failed to identify the correct secret.  all matching instrumentation's must use the same secret")
+		return pod, nil
+	} else {
+		err = pm.secretReplicator.ReplicateSecret(ctx, ns, pod, pm.operatorNamespace, licenseKeySecret)
+		if err != nil {
+			logger.Error(err, "failed to replicate secret")
+			return pod, nil
+		}
 	}
-	insts.Python = inst
 
-	if inst, err = pm.getInstrumentationInstance(ctx, ns, pod, annotationInjectDotNet); err != nil {
-		// we still allow the pod to be created, but we log a message to the operator's logs
-		logger.Error(err, "failed to select a New Relic Instrumentation instance for this pod")
-		return pod, err
+	return mutatedPod, nil
+}
+
+// GetLanguageInstrumentations is used to collect all instrumentations and validate that only a single insturmentation
+// exists for each language, and return them together
+func GetLanguageInstrumentations(instCandidates []*v1alpha2.Instrumentation) ([]*v1alpha2.Instrumentation, error) {
+	var instrumentations []*v1alpha2.Instrumentation
+	languages := map[string]*v1alpha2.Instrumentation{}
+	for _, candidate := range instCandidates {
+		if currentInst, ok := languages[candidate.Spec.Agent.Language]; ok {
+			if currentInst.Spec.Agent.Image != candidate.Spec.Agent.Image ||
+				!reflect.DeepEqual(currentInst.Spec.Agent.Env, candidate.Spec.Agent.Env) ||
+				!reflect.DeepEqual(currentInst.Spec.Agent.VolumeSizeLimit, candidate.Spec.Agent.VolumeSizeLimit) ||
+				!reflect.DeepEqual(currentInst.Spec.Agent.Resources, candidate.Spec.Agent.Resources) {
+				return nil, errMultipleInstancesPossible
```

Comment:
I wonder whether we should define if two Agents are deep equal in the `instrumentation_types.go` and just call the function here as `!currentInst.Spec.Agent.IsEqual(candidate.Spec.Agent)`

Reply:
That would clean that part up.  I like the approach, but wasn't sure since based on the name, you'd think `Agent.Language` would be included.  We could document that nuance though.

Reply:
updated